### PR TITLE
INT-3807: The `AmqpInboundGateway` improvements

### PR DESCRIPTION
**Pull request by artembilan**
JIRA: https://jira.spring.io/browse/INT-3807

### CODE REVIEW

**Comment by artembilan** [comment]
Also fixes: https://jira.spring.io/browse/INT-3810

File: spring-integration-amqp/src/main/java/org/springframework/integration/amqp/inbound/AmqpInboundGateway.java [hunk 1]
```diff
@@ -67,14 +87,17 @@ public AmqpInboundGateway(AbstractMessageListenerContainer listenerContainer) {
 						"the adapter needs to configure its own listener implementation.");
 		this.messageListenerContainer = listenerContainer;
 		this.messageListenerContainer.setAutoStartup(false);
-		this.amqpTemplate = new RabbitTemplate(this.messageListenerContainer.getConnectionFactory());
+		this.amqpTemplate = amqpTemplate;
+		this.amqpTemplateExplicitlySet = amqpTemplateExplicitlySet;
```
Comment (garyrussell) [inline spring-integration-amqp/src/main/java/org/springframework/integration/amqp/inbound/AmqpInboundGateway.java]: Need to assert not null, if explicitly set

File: spring-integration-amqp/src/main/java/org/springframework/integration/amqp/inbound/AmqpInboundGateway.java [hunk 2]
```diff
@@ -67,14 +87,17 @@ public AmqpInboundGateway(AbstractMessageListenerContainer listenerContainer) {
...
 	}
 
 
 	public void setMessageConverter(MessageConverter messageConverter) {
 		Assert.notNull(messageConverter, "MessageConverter must not be null");
 		this.amqpMessageConverter = messageConverter;
-		this.amqpTemplate.setMessageConverter(messageConverter);
+		if (!amqpTemplateExplicitlySet) {
+			((RabbitTemplate) this.amqpTemplate).setMessageConverter(messageConverter);
+		}
```
Comment (garyrussell) [inline spring-integration-amqp/src/main/java/org/springframework/integration/amqp/inbound/AmqpInboundGateway.java]: Need to explain in javadoc that the converter will not be applied to the template if it was injected.

File: spring-integration-amqp/src/main/java/org/springframework/integration/amqp/inbound/AmqpInboundGateway.java [hunk 3]
```diff
@@ -102,48 +139,54 @@ public void onMessage(Message message, Channel channel) {
...
+						}
+
+					};
+
+					if (replyTo != null) {
+						amqpTemplate.convertAndSend(replyTo.getExchangeName(), replyTo.getRoutingKey(),
+								reply.getPayload(), messagePostProcessor);
+					}
+					else {
+						amqpTemplate.convertAndSend(reply.getPayload(),	messagePostProcessor);
```
Comment (garyrussell) [inline spring-integration-amqp/src/main/java/org/springframework/integration/amqp/inbound/AmqpInboundGateway.java]: We should add some logic here - if not explicitly set, throw `IllegalStateException` - we know, for sure that the default template has no defaults for exchange/routing key.

We also need to clearly document the behavior - if you inject a template and don't set a default replyTo, the injected template's default exchange/routing key will be used to route replies.

File: spring-integration-amqp/src/main/resources/org/springframework/integration/amqp/config/spring-integration-amqp-4.2.xsd [hunk 4]
```diff
@@ -229,6 +229,29 @@
...
+							<xsd:documentation>
+								The AmqpTemplate bean reference to be used for sending replies.
+								Defaults to `RabbitTemplate` based on the provided `ConnectionFactory`.
+							</xsd:documentation>
+						</xsd:annotation>
+					</xsd:attribute>
+					<xsd:attribute name="default-reply-to">
+						<xsd:annotation>
+							<xsd:documentation>
+								The 'defaultReplyTo' address in the form '(exchange)/(routingKey)'
```
Comment (garyrussell) [inline spring-integration-amqp/src/main/resources/org/springframework/integration/amqp/config/spring-integration-amqp-4.2.xsd]: (or `(queueName)` - in which case the default exchange will be used with the queue name as the routing key.

File: spring-integration-amqp/src/test/java/org/springframework/integration/amqp/config/AmqpInboundGatewayParserTests.java [hunk 5]
```diff
@@ -88,6 +89,13 @@ public void verifyLifeCycle() {
 		assertEquals(Boolean.FALSE, TestUtils.getPropertyValue(gateway, "autoStartup"));
 		assertEquals(123, TestUtils.getPropertyValue(gateway, "phase"));
 		assertFalse(TestUtils.getPropertyValue(gateway, "messageListenerContainer.missingQueuesFatal", Boolean.class));
+		Object amqpTemplate = context.getBean("amqpTemplate");
+		assertSame(amqpTemplate, TestUtils.getPropertyValue(gateway, "amqpTemplate"));
+//		TODO AMQP-527
+//		assertEquals(new Address("fooExchange/barRoutingKey"), TestUtils.getPropertyValue(gateway, "defaultReplyTo"));
```
Comment (garyrussell) [inline spring-integration-amqp/src/test/java/org/springframework/integration/amqp/config/AmqpInboundGatewayParserTests.java]: This is merged now.

**Comment by artembilan** [comment]
Pushed

**Comment by garyrussell** [comment]
Polish here https://github.com/garyrussell/spring-integration/commit/78d1abcfd2f5884899c390824337d57194601417

If ok, please add to PR so Travis can verify.

**Comment by artembilan** [comment]
Pushed the polishing.
Waiting for the Travis result

**Comment by artembilan** [comment]
Travis has passed.
Just for an info if you'd like to merge...

**Comment by garyrussell** [comment]
Merged as 02565d04013d5a372c386f5c0d0f210325830d50

**Pull request closed without merging**